src/main/java/httpserver/HttpResponder.java

```diff
@@ -19,7 +20,7 @@ public HttpResponder(Parseable parser, Routeable router, Presentable presenter)
     @Override
     public void respond(Connection connection) throws IOException {
         String input = connection.read();
-        HttpRequest parsedInput = parser.parse(input);
+        Optional<HttpRequest> parsedInput = parser.parse(input);
         HttpResponse response = router.route(parsedInput);
```

Comment:
An alternative to passing in an optional to the router could be to map here in `respond`

```java
parser.parse(input)
  .map((request) -> router.route(request));
  .orElse(new BadRequest())
```

The advantage here is that the router doesn't have to know that the parser could return an optional. It only deals with `HttpRequest`s

Reply:
Doesn't it feel like a breach of responsibility to route the response within the respond method?

Reply:
That is a good question. Should the `Router` need to know about an `Optional`? Is there anything to route if there is no request?

I don't have the answer to either of these questions, so I think you should use your best judgment.